GPUP: Drawing paths spends time in adding segments during PathCG construction

kkinnunen-apple opened the pull request:
#### 7a9bc240e0303668a5dda399857603455c48de1b
<pre>
GPUP: Drawing paths spends time in adding segments during PathCG construction
<a href="https://bugs.webkit.org/show_bug.cgi?id=292595">https://bugs.webkit.org/show_bug.cgi?id=292595</a>
<a href="https://rdar.apple.com/150750039">rdar://150750039</a>

Reviewed by Simon Fraser.

Avoid calling PathCG virtual functions when adding path segments to
the platform path. Instead, call the target CG functions directly.

* Source/WebCore/WebCore.order:
* Source/WebCore/platform/graphics/Path.cpp:
(WebCore::Path::ensurePlatformPathImpl):
* Source/WebCore/platform/graphics/PathImpl.cpp:
(WebCore::PathImpl::beziersForRoundedRect):
(WebCore::PathImpl::addBeziersForRoundedRect): Deleted.
* Source/WebCore/platform/graphics/PathImpl.h:
* Source/WebCore/platform/graphics/cairo/PathCairo.cpp:
(WebCore::PathCairo::create):
(WebCore::PathCairo::add):
* Source/WebCore/platform/graphics/cairo/PathCairo.h:
* Source/WebCore/platform/graphics/cg/PathCG.cpp:
(WebCore::addToCGPath):
(WebCore::addToCGContextPath):
(WebCore::addUnevenCornersRoundedRect):
(WebCore::PathCG::create):
(WebCore::PathCG::PathCG):
(WebCore::PathCG::definitelyEqual const):
(WebCore::PathCG::copy const):
(WebCore::PathCG::platformPath const):
(WebCore::PathCG::ensureMutablePlatformPath):
(WebCore::PathCG::add):
* Source/WebCore/platform/graphics/cg/PathCG.h:
* Source/WebCore/platform/graphics/skia/PathSkia.cpp:
(WebCore::PathSkia::create):
(WebCore::PathSkia::add):
* Source/WebCore/platform/graphics/skia/PathSkia.h:

Canonical link: <a href="https://commits.webkit.org/294622@main">https://commits.webkit.org/294622@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/453e00bab96d38831ae5ab7fa3ac923e26462b3d

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/102455 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/22124 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/12438 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/107616 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/53092 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/104494 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/22432 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/30630 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/77948 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/34941 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/105461 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/17377 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/92481 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/58285 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/17214 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/10509 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/52449 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/87031 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/10580 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/109991 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/29587 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/21817 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/86930 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/29951 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/88677 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/86523 "Found 102 new API test failures: /WebKitGTK/TestUIClient:/webkit/WebKitWebView/javascript-dialogs, /TestWebKit:WebKit.DOMWindowExtensionCrashOnReload, /TestWebKit:WebKit.PageLoadBasic, /TestWebKit:WebKit.NewFirstVisuallyNonEmptyLayoutFails, /WebKitGTK/TestInputMethodContext:/webkit/WebKitInputMethodContext/content-type, /TestWebKit:WebKit.InjectedBundleBasic, /TestWebKit:WebKit.PreventEmptyUserAgent, /WebKitGTK/TestWebKitPolicyClient:/webkit/WebKitPolicyClient/autoplay-policy, /WebKitGTK/TestInputMethodContext:/webkit/WebKitInputMethodContext/surrounding, /WebKitGTK/TestWebViewEditor:/webkit/WebKitWebView/select-all/non-editable ... (failure)") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/22015 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/31348 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/9069 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/23832 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/29515 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/34818 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/29326 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/32649 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/30885 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

smfr commented on Source/WebCore/platform/graphics/PathImpl.cpp:
```diff
@@ -50,37 +50,41 @@ void PathImpl::addBeziersForRoundedRect(const FloatRoundedRect& roundedRect)
 
-    add(PathCloseSubpath { });
+    segments.append(PathSegment(PathCloseSubpath { }));
+    return segments;
```
Maybe assert that segments.size() == 9 here in case someone copy-pastes this code for continuous rounded rects.

kkinnunen-apple committed:
GPUP: Drawing paths spends time in adding segments during PathCG construction
https://bugs.webkit.org/show_bug.cgi?id=292595
rdar://150750039

Reviewed by Simon Fraser.

Avoid calling PathCG virtual functions when adding path segments to
the platform path. Instead, call the target CG functions directly.

* Source/WebCore/WebCore.order:
* Source/WebCore/platform/graphics/Path.cpp:
(WebCore::Path::ensurePlatformPathImpl):
* Source/WebCore/platform/graphics/PathImpl.cpp:
(WebCore::PathImpl::beziersForRoundedRect):
(WebCore::PathImpl::addBeziersForRoundedRect): Deleted.
* Source/WebCore/platform/graphics/PathImpl.h:
* Source/WebCore/platform/graphics/cairo/PathCairo.cpp:
(WebCore::PathCairo::create):
(WebCore::PathCairo::add):
* Source/WebCore/platform/graphics/cairo/PathCairo.h:
* Source/WebCore/platform/graphics/cg/PathCG.cpp:
(WebCore::addToCGPath):
(WebCore::addToCGContextPath):
(WebCore::addUnevenCornersRoundedRect):
(WebCore::PathCG::create):
(WebCore::PathCG::PathCG):
(WebCore::PathCG::definitelyEqual const):
(WebCore::PathCG::copy const):
(WebCore::PathCG::platformPath const):
(WebCore::PathCG::ensureMutablePlatformPath):
(WebCore::PathCG::add):
* Source/WebCore/platform/graphics/cg/PathCG.h:
* Source/WebCore/platform/graphics/skia/PathSkia.cpp:
(WebCore::PathSkia::create):
(WebCore::PathSkia::add):
* Source/WebCore/platform/graphics/skia/PathSkia.h:

Canonical link: https://commits.webkit.org/294622@main
